craft_parts/state_manager/state_manager.py

```diff
@@ -0,0 +1,152 @@
+        """
+        return self.serial > other.serial
+
+
+class _EphemeralStates:
+    """A dictionary-backed simple database manager for wrapped states."""
+
+    def __init__(self):
+        self._state: Dict[str, Dict[Step, _StateWrapper]] = {}
+        self._serial_gen = itertools.count(1)
+
+    def wrap_state(self, state: StepState, step_updated: bool = False) -> _StateWrapper:
+        """Add metadata to step state.
+
+        :param state: The part state to store.
+        :param step_updated: Whether this state was updated after an
+            outdated report.
+
+        :return: The wrapped state with additional metadata.
+        """
+        stw = _StateWrapper(
+            state, serial=next(self._serial_gen), step_updated=step_updated
+        )
+        return stw
+
+    def set(
+        self, *, part_name: str, step: Step, state: Optional[_StateWrapper]
+    ) -> None:
+        """Set a state for a given part and step.
+
+        :param part_name: The name of the part corresponding to the state
+            to be set.
+        :param step: The step corresponding to the state to be set.
+        :param state: The state to assign to the part name and step.
+        """
+        if not state:
+            self.remove(part_name=part_name, step=step)
+            return
+
+        if part_name not in self._state:
+            self._state[part_name] = dict()
+
+        self._state[part_name][step] = state
+
+    def get(self, *, part_name: str, step: Step) -> Optional[_StateWrapper]:
+        """Retrieve the state for a given part and step.
+
+        :param part_name: The name of the part corresponding to the state
+            to be retrieved.
+        :param step: The step corresponding to the state to be retrieved.
+
+        :return: The wrapped state assigned to the part name and step.
+        """
+        if self.test(part_name=part_name, step=step):
+            return self._state[part_name][step]
+        return None
+
+    def test(self, *, part_name: str, step: Step) -> bool:
+        """Verify if there is a state defined for a given part and step.
+
+        :param part_name: The name of the part corresponding to the state
+            to be tested.
+        :param step: The step corresponding to the state to be tested.
+
+        :return: Whether a state is defined for the part name and step.
+        """
+        if part_name not in self._state:
+            return False
+        return step in self._state[part_name]
+
+    def remove(self, *, part_name: str, step: Step) -> None:
+        """Remove the state for a given part and step.
+
+        :param part_name: The name of the part corresponding to the state
+            to be removed.
+        :param step: The step corresponding to the state to be removed.
+        """
+        if part_name in self._state:
+            self._state[part_name].pop(step, None)
+
+    def rewrap(self, *, part_name: str, step: Step, step_updated: bool = False) -> None:
```

Comment:
```suggestion
    def update_part_step_metadata(self, *, part_name: str, step: Step, step_updated: bool = False) -> None:
```

Reply:
It was called `update_metadata` initially, but I renamed it to avoid confusion between this update and the `updated_state` flag.